### ISSUE TITLE
How print a very long single page PDF

### ISSUE THREAD

**Issue by zuoliang0** [sumatrapdfreader/sumatrapdf#4736]
**Is your feature request related to a problem? Please describe.**
 I have a very long A4 width pdf. My printer is a dot matrix printer that can automatically feed paper. I tried to use this software to print it, but I only got one A4 width page. 

**Describe the solution you'd like**
What I hope is to be able to achieve something like custom paper, so that the content of this page can be printed out completely (without scaling)
Is it possible to customize a paper size by read the width and height of the PDF when printing?

**Describe alternatives you've considered**
My current alternative is to render the PDF into A4 width and height, divided into many pages, which can be printed at present, but because each page has a little blank space at the end, the printed content will be interrupted (my PDF is a long table)

**Additional context**
[print_1735184594331.pdf](https://github.com/user-attachments/files/18251985/print_1735184594331.pdf) test pdf file

**Comment by GitHubRulesOK** [comment]
***but I only got one A4 width page.*** is how the default printer is defined. 

SumatraPDF does not alter the source page dimensions nor describe how to be printed those are all set by the system but there are constraints. SumatraPDF will simply honour the system settings at the time of print handover as pixels so if the system requires A4 pages or requires USA Letter then SumatraPDF uses those page defined pixels units.

Printers have a limited length of paper sizes so a GhostScript printer may easily be limited to a nominal 2 m or 5 m but SumatraPDF can use those length no problem.

If you are using PDF to be printed to a non default size then GhostScript PDL is probably the best command line method for a heads up use it is more restricted. I have shown SumatraPDF can work with pages of up to 200" nominal length but dont recommend it be done that way. https://stackoverflow.com/a/78321355/10802527

**Comment by zuoliang0** [comment]
I still don't quite understand, what do I need to do to print this file completely?

I just tested using GhostScript to print this file directly, and it can only print the first page (A4 length)
Then I tested manually specifying the width and height of the device, and the printing prompt would be an error
``` sh
gswin32 -dNOPAUSE -dBATCH -sDEVICE=mswinpr2 -sOutputFile="%printer%EPSON LQ-82KF ESC/P2 (副本 1)" -dDEVICEWIDTHPOINTS 794 -dDEVICEHEIGHTPOINTS 6060 output1.pdf
```
![image](https://github.com/user-attachments/assets/7f1eab2e-7d0d-4e0b-93d3-7a28740d8179)

I tried many solutions, but it still doesn't work. Am I trying the wrong direction?

**Comment by GitHubRulesOK** [comment]
Command line printers expect to be run in a system command console you seem to be trying to print from within some windows  wrapper application ?

Try your commands in the windows command console before move to any form of embedment. use gswin32C.exe

PDF is unitless but has a maximum length so you need to debug what units you are attempting.
also devices may need to be physically configured to use rolls of paper not sheets. that looks like a sheet printer.

**Comment by zuoliang0** [comment]
Thank you very much for your patience.

Yes, we are developing a program for printing HTML to a printer. At first we tried to print directly, but encountered paging problems. Then we tried to make the HTML into a long single-page PDF, but it seems that software like SumatraPDF or pdftopinter can not support direct printing.

I used mutool to crop the file according to your method. From the PDF point of view, there is indeed no upper and lower margins, but when it is printed out through the printer, a blank line will still be generated after printing a page.

I think this blank line is because of the natural behavior of the printer, which can't be avoided by the way we cut pdf.

I feel powerless.

**Comment by GitHubRulesOK** [comment]
PDF in effect is borderless i.e. does not have any "margins" as such 
So if you have a page of 3.5 x A4 long you would need to make that an additional white space at end of last page to be half a void long this is relatively easy but slightly counter intuitive since that would be a negative half page below the bottom left origin that a PDF uses.

cpdf has the ability to modify page shapes to suit 4 x A4 high before chop and had shown such example in other forums but was deleted ! 

If you simply need images of pages to send to a printer since  ink can be pixels not vectors, it is perhaps simpler to use Magick and here I described a similar case. https://stackoverflow.com/a/78615507/10802527 the accepted answer (below mine) may or may not be better

**Comment by zuoliang0** [comment]
It depends. This problem has nothing to do with cropping. I have provided a cropped PDF here. You can see that there are no borders. However, after printing a page through the printer, the printer will still retain 2 lines of blank space.
[out.pdf](https://github.com/user-attachments/files/18278676/out.pdf)
``` sh
mutool -x 1 -y 5 out.pdf
```
I'm trying other methods, maybe I don't need to convert HTML to PDF before printing

**Comment by GitHubRulesOK** [comment]
Hmmm several problems with such tables. 

Microsoft Edge says page size is "280 × 428 mm (portrait)" I know that is clearly wrong so the pages are physically 210 x 321 mm ? which is not European nor American printing paper.

A bigger problem is that it is "tabular" so no easy places to cut unless you aim for a border between cells.

The one tool I know can split with such precision (but needs some trial and error to get perfect) is cpdf -chop-h where you split and split again until all pages are done. 

Recently gave an example elsewhere of vertical slicing with 2 precision cuts but it worked https://stackoverflow.com/a/79314388/10802527

**Comment by GitHubRulesOK** [comment]
OK forget that suggestion it is exceptionally slow in chopping a very long page such as that, and would be hard to decide where to split. We call this an XY problem where trying to fix a bad input siply is not worth it when the problem lies with the generation of such un-paged format. Page Breaks should be done at the source.
[output.pdf](https://github.com/user-attachments/files/18279155/output.pdf)

```
cpdf -chop-h 3399 print_1735184594331.pdf -o 2bits.pdf
cpdf -chop-h 2576 2bits.pdf 2 -o 3bits.pdf
cpdf -chop-h 1760 3bits.pdf 3 -o 4bits.pdf
cpdf -chop-h 935 4bits.pdf 4 -o 5bits.pdf
cpdf -chop-h 100 5bits.pdf 5 -o output.pdf
del *bits.pdf
```

**Comment by zuoliang0** [comment]
Yes, not only is it troublesome to cut, but the PDF after paging will still have a blank line when printed. I suspect it is a problem with the printer.

**Comment by zuoliang0** [comment]
Thanks to @GitHubRulesOK , this problem seems not something that SumatraPDF can and should solve.

**Comment by GitHubRulesOK** [comment]
@zuoliang0 
There are problems with the way Windows manages printing So SumatraPDF **can** send longer prints IF the driver allows for custom formats as seen in this double high printout [xodo.pdf](https://github.com/user-attachments/files/18287349/xodo.pdf)

HOWEVER that is still a fixed Windows FORMat that needs to be defined in Windows Print Management then selected IF alowed by the print driver. It cannot be longer or shorter it is still a "FIXED FORMAT" not a continuous one which you need to resolve your problem

**Comment by zuoliang0** [comment]
@GitHubRulesOK Thank you very much. 
I have some new research progress. 

The printer I use is an Epson dot matrix printer. After reading the manual, I found that I can send a control command such as ** ECS 0 ** to close the blank line at the end of the page. Then, it seems that even with paging, it can print a similar effect of continuous printing. However, it seems that such control commands cannot be transmitted when printing directly through the Windows driver (or my technology is not enough). 

Because I hope to develop a general program, if I use a printer-specific control command to close the trailing space, then once the printer brand is changed, it will cause new problems, so I have not done any further research yet.

**Comment by GitHubRulesOK** [comment]
@zuoliang0 
Thanks for the feedback but when designing custom direct printouts they have to be in the printers own language as designed for the different platforms So there really is no generic way to print to different printers (Thus windows uses a Common Page system as a means to channel applications through a common "Page" spooler but the  core print contents can still vary per driver.

Page Description Languages are developed by Printer manufacturers So there is 
Zebra ZPL (plus others) 
HP has HPGL  PCL and many others 
Apple PostScript etc.
Epson Esc/POS Esc/P1 Esc/P2 (others ?) 

There are several dozen manufactures with multiple languages for different printers. Thus a nightmare if you dont use standard pages which too vary depending on user locale.
